В цьому домашньому завданні продовжуємо працювати з набором даних supermarket_sales.csv.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну df типу pandas.DataFrame.


In [50]:
import pandas as pd
import numpy as np

In [51]:
import os
os.listdir('../data')

['.ipynb_checkpoints',
 'sample_submission.csv',
 'supermarket_sales.csv',
 'train.csv']

In [52]:
data_path = '../data/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [53]:
total_gross_income = df.groupby('Branch')['gross income'].sum()
total_gross_income

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.
Знайдіть відповідь за допомогою функціоналу Pandas.

In [54]:
most_profitable_city = df.groupby(['Branch','City'])['gross income'].sum().round(3)
most_profitable_city

Branch  City     
A       Yangon       5057.160
B       Mandalay     5057.032
C       Naypyitaw    5265.176
Name: gross income, dtype: float64

In [55]:
economic_results = df.groupby(['Branch','City']).agg(
    total_revenue=('Total', 'sum'),
    avg_check=('Total', 'mean'),
    avg_gross_income=('gross income', 'mean'),
    total_transactions=('Invoice ID', 'count')).round(3)
economic_results 
    

,,total_revenue,avg_check,avg_gross_income,total_transactions
Branch,City,,,,
A,Yangon,106200.370,312.354,14.874,340
B,Mandalay,106197.672,319.873,15.232,332
C,Naypyitaw,110568.706,337.100,16.052,328


Супермаркет у місті Naypyitaw має найвищі показники в наступних категоріях: 
- gross income(5265.176),
- avg_gross_income(16.052),
- total_revenue(110568.706),
- avg_check(337.100),

На основі цих даних ми бачимо, що цей магазин має більший середній чек, хоча кількість транзакцій є меншою, ніж у решти супермаркетів. Це означає, що купівельна спроможність цих покупців вища, ніж в інших містах. Відповідно і загальна виручка більша. Тому потенційно доцільно відкрити додатковий магазин, бо є платоспроможний попит в цьому регіоні.
    
    

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (Branch) за різними категоріями товарів. Запишіть таблицю в змінну invoices_by_category і виведіть змінну на екран. Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [56]:
invoices_by_category = pd.pivot_table(df, index='Branch', columns='Product line', values='Invoice ID', aggfunc='count')
invoices_by_category 

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


In [57]:
df.head()

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3


3.2. Викристовуючи змінну invoices_by_category дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (Branch) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (Electronic accessories)?

In [58]:
invoices_by_category['Electronic accessories'].sort_values(ascending=False).head(1)

Branch
A    60
Name: Electronic accessories, dtype: int64

4-6. Творче завдання на розвиток аналітичного мислення

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.

Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌

In [59]:
df.head()

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3


Гіпотеза 1.
    Які TOP-3 категорії товарів є найпопулярнішими за виторгом загалом, і як розподіляються покупки в цих категоріях між чоловіками та жінками у розрізі кожного міста?

In [83]:
top_categories = df.groupby('Product line')['Total'].sum().sort_values(ascending=False).nlargest(3).index
top_categories
    


Index(['Food and beverages', 'Sports and travel', 'Electronic accessories'], dtype='object', name='Product line')

In [84]:
pivot_top = df.pivot_table(index=['Gender', 'City'], columns='Product line', values='Total', aggfunc='sum').round(2)
pivot_top

Product line      Electronic accessories  Fashion accessories  \
Gender City                                                     
Female Mandalay                  8166.51              9081.49   
       Naypyitaw                 8969.22             11519.69   
       Yangon                    9966.30              9836.22   
Male   Mandalay                  8884.93              7331.82   
       Naypyitaw                 9999.76             10040.38   
       Yangon                    8350.82              6496.29   

Product line      Food and beverages  Health and beauty  Home and lifestyle  \
Gender City                                                                   
Female Mandalay             10510.00            6400.35             9557.13   
       Naypyitaw            15661.16            6445.74             7842.93   
       Yangon                6999.76            5714.90            12636.81   
Male   Mandalay              4704.89           13580.31             7992.03   
       Naypyitaw             8105.70           10169.59             6052.62   
       Yangon               10163.34            6882.86             9780.38   

Product line      Sports and travel  
Gender City                          
Female Mandalay             9212.82  
       Naypyitaw           11246.73  
       Yangon               8115.18  
Male   Mandalay            10775.38  
       Naypyitaw            4515.20  
       Yangon              11257.52

In [86]:
pivot_top3 = pivot_top[top_categories].round(2)
pivot_top3

Product line      Food and beverages  Sports and travel  \
Gender City                                               
Female Mandalay             10510.00            9212.82   
       Naypyitaw            15661.16           11246.73   
       Yangon                6999.76            8115.18   
Male   Mandalay              4704.89           10775.38   
       Naypyitaw             8105.70            4515.20   
       Yangon               10163.34           11257.52   

Product line      Electronic accessories  
Gender City                               
Female Mandalay                  8166.51  
       Naypyitaw                 8969.22  
       Yangon                    9966.30  
Male   Mandalay                  8884.93  
       Naypyitaw                 9999.76  
       Yangon                    8350.82

**Food and beverages**:
У місті Naypyitaw жінки витрачають $15 661.16 проти $8 105.70  у чоловіків. У місті Yangon ситуація протилежна — чоловіки купують більше ($10 163.34 проти $6 999.76).

**Sports and travel**: У місті Naypyitaw цю категорію активно купують жінки ($11 246.73 проти $4 515.20), а у Yangon та Mandalay виторг вищий серед чоловіків ($10775.38 проти $9212.82)

**Electronic accessories**: У місті Yangon жінки витрачають більше ($9 966.30), а в Naypyitaw — чоловіки ($9 999.76).

**Рекомендації для бізнесу**:
1.Запускати таргетований маркетинг для продуктової групи (Food & beverages) у Yangon з орієнтацією на жінок, а в Mandalay — на чоловіків.
2.Сформувати акційні пропозиції відповідно до статевих переваг у кожному регіоні
3.Попрацювати над мерчендайзінгом спорт товарів у місті Yangon для жінок, та в Naypyitaw для чоловіків.

Гіпотеза 2.
    Чи існує взаємозв'язок між обраним способом оплати (Cash, Credit card, Ewallet), середньою кількістю товарів у чеку (Quantity) та середнім значенням рейтингу задоволеності клієнта (Rating)?

In [109]:
deversification_by_payment = df.groupby('Payment').agg(
    avg_quantity=('Quantity', 'mean'),
    avg_rating=('Rating', 'mean'),
    total_transactions=('Quantity', 'count')
).round(2)

In [110]:
deversification_by_payment

,avg_quantity,avg_rating,total_transactions
Payment,,,
Cash,5.51,6.97,344
Credit card,5.54,7.00,311
Ewallet,5.48,6.95,345


Спосіб оплати практично не впливає ані на середнє наповнення чека товарів (~5.5 шт.), ані на рівень задоволеності клієнтів (~7.0 балів).

**Рекомендації для бізнесу**
    Попри те, що **Credit card** має найвищий рейтинг **(7.00)**, цей метод є найменш популярним за кількістю транзакцій **(311 проти ~345 у Cash та Ewallet)**. Варто з'ясувати причини нижчого використання кредиток (наприклад, термінали довше обробляють операції або є відсоток відмов/комісій) та стимулювати цей канал через партнерські кешбеки чи програми лояльності.

Гіпотеза 3.
   Чи існує різниця у середній сумі податку/націнки (Tax 5%) між транзакціями, здійсненими через Ewallet, Credit card та Cash?

In [114]:
tax_inf = df.groupby('Payment')['Tax 5%'].agg(['mean', 'median', 'sum']).round(3)
tax_inf

,mean,median,sum
Payment,,,
Cash,15.532,12.522,5343.170
Credit card,15.429,11.409,4798.432
Ewallet,15.182,12.080,5237.767


Середній податок для всіх форм оплати є практично однаковим ($15.182 - 15.532). Це означає, що середня вартість чека не залежить від того, як клієнт здійснив оплату.
Медіана для оплат кредитками становить $11.409 і є меншою в порівнянні з іншими методами оплати. Це говорить про те, що кредитними картками частіше оплачували дрібніпокупки.
Сума податку для Credit card суттєво нижча і становить $4798.432. Вона зумовлена виключно меншою кількістю проведених транзакцій (311 проти 344-345).

**Рекомендації для бізнесу**
    Оскільки середня вартість транзакцій однакова для всіх типів платежів, бізнесу варто мотивувати клієнтів частіше використовувати Credit card (наприклад, через кешбек або знижки від банків-партнерів), щоб зрівняти кількість транзакцій з Cash та Ewallet.